File: app/src/main/java/com/locallinkonline/stcatherineschool/rest/tasks/GetNewScheduleTask.java
package com.locallinkonline.stcatherineschool.rest.tasks;

import com.locallinkonline.stcatherineschool.rest.model.Schedule;
import com.locallinkonline.stcatherineschool.room.dao.SchoolScheduleDao;
import com.locallinkonline.stcatherineschool.room.db.StCatherineDatabase;
import com.locallinkonline.stcatherineschool.room.entity.ScheduleEntity;

import java.io.IOException;
import java.util.List;

import retrofit2.Call;
import retrofit2.Response;

public class GetNewScheduleTask extends StCatherineTask<String> {

    SchoolScheduleDao dao;

    public GetNewScheduleTask(StCatherineDatabase db, String url) {
        super(db, url);
        this.dao = db.schoolScheduleDao();
    }

    @Override
    protected Void doInBackground(String... args) {

        String identifier = args != null && args.length > 0 && args[0] != null ? args[0] : "school-schedule";
        String type = args != null && args.length > 0 ? args[1] : null;

        Call<Schedule> call;

        if("Homework".equals(type)) {
            call = stCatherineApi.getHomeworkSchedule(identifier);
        } else if ("Sports".equals(type)) {
            call = stCatherineApi.getActivitySchedule(identifier);
        } else {
            call = stCatherineApi.getSchoolSchedule();
        }

        try {
            Response<Schedule> response = call.execute();

            List<ScheduleEntity> entities = response.body().getSchedule();

            for(int i = 0; i < entities.size(); i++){
                entities.get(i).setScheduleType(identifier);
            }

            if(entities.size() > 0) {
                dao.insert(entities);
            }

        } catch (IOException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return null;
    }
}

File: app/src/main/java/com/locallinkonline/stcatherineschool/rest/tasks/GetNewScheduleTask.kt
package com.locallinkonline.stcatherineschool.rest.tasks

import com.locallinkonline.stcatherineschool.rest.model.Schedule
import com.locallinkonline.stcatherineschool.room.dao.SchoolScheduleDao
import com.locallinkonline.stcatherineschool.room.db.StCatherineDatabase
import com.locallinkonline.stcatherineschool.room.entity.ScheduleEntity

import java.io.IOException

import retrofit2.Call
import retrofit2.Response

class GetNewScheduleTask(db: StCatherineDatabase, url: String) : StCatherineTask<String>(db, url) {

    internal var dao: SchoolScheduleDao = db.schoolScheduleDao()

    override fun doInBackground(vararg args: String): Void? {

        val identifier = if (args.size > 0) args[0] else "school-schedule"
        val type = if (args.size > 1) args[1] else "SchoolSchedule"

        val call: Call<Schedule>

        if ("Homework" == type) {
            call = stCatherineApi.getHomeworkSchedule(identifier)
        } else if ("Sports" == type) {
            call = stCatherineApi.getActivitySchedule(identifier)
        } else {
            call = stCatherineApi.schoolSchedule
        }

        try {
            val response = call.execute()

            val entities = response.body()!!.schedule

            for (i in entities!!.indices) {
                entities[i].scheduleType = identifier
            }

            if (entities.size > 0) {
                dao.updateScheduleItems(entities)
            }

        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return null
    }
}
